README.md
# Pexels for Craft CMS 4

Pexels is a CraftCMS plugin that seamlessly integrates Pexels.com image stock library search functionality into your CraftCMS asset management. With this plugin, you can easily search, download, and import images from the Pexels library directly into your CraftCMS project.

## Features

- Adds a "Search Pexels" button to asset selection windows
- Automatically downloads and saves selected Pexels images to the server
- Streamlines image search and import workflow within CraftCMS 
- Option to choose the search query language (supported languages by Pexels)
- Option to configure the number of results displayed per page (in settings)
- Option to search for images by orientation: portrait, landscape, or square 
- Option to specify the minimum photo size for search results (in megapixels)
- Option to search for photos based on their predominant color

## Requirements

- Pexels.com Account (Free): You need to have a Pexels account and a Pexels API key, which can be found at [https://www.pexels.com/api/](https://www.pexels.com/api/)
- Craft CMS 4.0.0 or later

## Installation

1. Download the Pexels plugin from plugins.craftcms.com
2. Activate the plugin in the CraftCMS Control Panel
3. Navigate to the plugin settings and enter your Pexels API key
4. Configure the default download settings in plugin settings

## Usage

1. Navigate to an asset selection window in CraftCMS (e.g. when editing an entry)
2. Click the "Search Pexels" button to open the Pexels search modal
3. Search for images and select the ones you want to download
4. Click the "Download Selected" button to automatically download and save the images to your server

## Support

For any issues or questions, you can reach us by email info@convergine.com or by opening an issue on GitHub.

**Disclaimer:** By default you get 200 API requests per hour and 20,000 requests per month. If you need more, you can contact Pexels.com support to get a higher limit. Convergine is not affiliated with Pexels.com. 


CHANGELOG.md
# Pexels - Changelog

All notable changes to this project will be documented in this file.

## 1.0.0 - 2023-04-20
### Added
- Initial release


src/controllers/PexelsController.php
<?php
namespace convergine\pexels\controllers;

use Craft;
use craft\elements\Asset;
use craft\web\Controller;
use convergine\pexels\PexelsPlugin;

class PexelsController extends Controller
{

    public function actionSearchPexels(){
        /* search Pexels action */
        $this->requireAcceptsJson();

        $request = Craft::$app->request;
        $query = $request->getRequiredParam('query');
        $page = max((int)$request->getParam('page', 1), 1);

        $params = [
            'orientation' => $request->getParam('orientation', ''),
            'size' => $request->getParam('size', ''),
            'color' => $request->getParam('color', '')
        ];

        $url = $this->_apiUrlHelper($query, $page, $params['orientation'], $params['size'], $params['color']);

        $settings = PexelsPlugin::$plugin->settings;
        $apiKey = $settings->apiKey;
        $context = stream_context_create([
            'http' => ['header' => "Authorization: {$apiKey}\r\n"]
        ]);

        $pexelsData = @json_decode(@file_get_contents($url, false, $context), true);
        $pexelsData = array_merge($pexelsData, $params, ['query' => $query, 'page' => $page]);

        $success = isset($pexelsData['photos']);

        return $this->asJson(['success' => $success, 'data' => $pexelsData]);
    }


    public function actionDownload() {
		/* download Pexels photo action */

		$this->requireAcceptsJson();

		$url = Craft::$app->request->getRequiredParam('url');
		$target = Craft::$app->request->getRequiredParam('target');

        $folderId = $volumeId = null;
        $uploadLocation = PexelsPlugin::getInstance()->getSettings()->uploadLocation;

        if ($uploadLocation) {
            foreach (explode('/', $uploadLocation) as $location) {
                $locParts = explode(':', $location);
                if ($locParts[0] == 'volume' || $locParts[0] == 'volumeId') {
                    $volumeId = $locParts[1] ?? null;
                }
                if ($locParts[0] == 'folder' || $locParts[0] == 'folderId') {
                    $folderId = $locParts[1] ?? null;
                }
            }
        }

		if ($folderId) {

			/* find the folder, either by id or by uid */
            $folder = Craft::$app->assets->findFolder([(is_numeric($folderId) ? 'id' : 'uid') => $folderId]);
            if (!$folder) {
                throw new \Error('Folder not found: ' . $folderId);
            }
            $volume = $folder->volume;

		} elseif ($volumeId) {

			/* find the volume */
			$fn = 'getVolumeBy' . (is_numeric($volumeId) ? 'Id' : 'Uid');
			$volume = Craft::$app->volumes->$fn($volumeId);
            if (!$volume) {
                throw new \Error('Volume not found: ' . $volumeId);
            }

            /* find the base folder */
            $folder = Craft::$app->assets->findFolder(['volumeId' => $volume->id, 'parentId' => ':empty:']);

            if (!$folder) {
                throw new \Error('Folder not found for volume: ' . $volumeId);
            }

		} else {
			throw new \Error('Unknown upload target: ' . $target);
		}

		/* check user's permission */
        if (!Craft::$app->user->checkPermission('saveAssets:' . $volume->uid)) {
            $this->requireAdmin();
        }

        /* prepare the photo url */
        if (!preg_match('/^https?\:\/\/[^\/]*pexels.com\//', $url)) {
            throw new \Error('This is not a Pexels URL: ' . $url);
        }

        $filename = preg_replace('/^.*\//', '', $url);

        if (!$filename) {
            throw new \Error('Could not get filename from URL: ' . $url);
        }

        $cacheDir = PexelsPlugin::tempDirectory();
        $apiKey = PexelsPlugin::$plugin->settings->apiKey;

		/* use curl to download the photo, change time limit to avoid timeouts */
        set_time_limit(600);
		$fp = fopen($cacheDir . DIRECTORY_SEPARATOR . $filename, 'w+');
		$ch = curl_init($url);

        /* add the authorization header with the API key */
        $headers = ["Authorization: {$apiKey}"];
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
		curl_setopt($ch, CURLOPT_TIMEOUT, 600);
		curl_setopt($ch, CURLOPT_FILE, $fp);
		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
		curl_exec($ch);
		curl_close($ch);
		fclose($fp);

		/* save the photo */
		$asset = new Asset();
		$asset->tempFilePath = $cacheDir . DIRECTORY_SEPARATOR . $filename;
        $asset->filename = $filename;
		$asset->newFolderId = $folder->id;
		$asset->setVolumeId($volume->id);
		$asset->uploaderId = Craft::$app->user->id;
		$asset->avoidFilenameConflicts = true;
		$asset->setScenario(Asset::SCENARIO_CREATE);

		$success = Craft::$app->elements->saveElement($asset);

		return $this->asJson(['success' => $success, 'assetId' => $asset->id, 'errors' => $asset->getErrors()]);

	}

    private function _apiUrlHelper($query, $page = 1, $orientation = '', $size = '', $color = '') {
        /* prep the api query URL */
        $settings = PexelsPlugin::$plugin->settings;
        $perPage = $settings->picsPerPage;
        $langLocale = $settings->langLocale;
        return 'https://api.pexels.com/v1/search?query=' . urlencode($query) . '&per_page=' . $perPage. '&locale=' . $langLocale . '&page=' . $page . '&orientation=' . $orientation. '&size=' . $size. '&color=' . $color;
    }

}


src/translations/en/convergine-pexels.php
<?php return [
    /* Settings Page */
    'Settings' => 'Settings',
    'Please Select' => 'Please Select',
    'pexelsApiNote' => 'Please note, the Pexels API comes with a default rate limit of 200 requests per hour and 20,000 requests per month.<br/> To request a higher limit, you should reach out to Pexels directly (api@pexels.com). If you adhere to Pexels API terms, you may be eligible to receive unlimited requests for free. <br/>',
	'missingApiKey' => 'API key is missing. Add your Pexels API key and click Save button.',
	'missingUploadLocation' => 'Upload location not set. Select location from dropdown and click Save button.',
    'labelApiKey' => 'Pexels API key',
    'instructionsApiKey' => 'Get your API key <a href="{url}" target="_blank">here</a> if you don\'t have one.',
    'labelUploadLocation' => 'Assets Upload Location',
    'instructionsUploadLocation' => 'Select existing Volume or Folder. If none available - create one first.',
    'convergineHelp' => 'Convergine is not affiliated with Pexels.<br/>If you need help, please open issue on github or check out <a href="{url}" target="_blank">docs</a>.',
    'labelPicsPerPage' => 'Photos Per Page',
    'instructionsPicsPerPage' => 'Number of photos to display per page in search results.',
    'labelLangLocale' => 'Search Language',
    'instructionsLangLocale' => 'Select the language in which you will be searching the Pexels\' photos database.',
    /* Modal Popup */
    'Any Orientation' => 'Any Orientation',
    'Minimum Size' => 'Minimum Size',
    'Landscape' => 'Landscape',
    'Portrait' => 'Portrait',
    'Square' => 'Square',
    'Small (4MP)' => 'Small (4MP)',
    'Medium (12MP)' => 'Medium (12MP)',
    'Large (24MP)' => 'Large (24MP)',
    'Predominant Color' => 'Predominant Color',
    'Red' => 'Red',
    'Orange' => 'Orange',
    'Yellow' => 'Yellow',
    'Green' => 'Green',
    'Turquoise' => 'Turquoise',
    'Blue' => 'Blue',
    'Violet' => 'Violet',
    'Pink' => 'Pink',
    'Brown' => 'Brown',
    'Black' => 'Black',
    'Gray' => 'Gray',
    'White' => 'White',
    'Download Selected' => 'Download Selected',
    'Downloading, please wait.' => 'Downloading, please wait.',
    'placeholderString' => 'Use the search box above to search Pexels pictures database.',
    'An error occurred while loading Pexels data' => 'An error occurred while loading Pexels data',
    '0 photos found for "{query}"' => '0 photos found for "{query}"',
    'Previous {picPerPage}' => 'Previous {picPerPage}',
    'Next {picPerPage}' => 'Next {picPerPage}',
    'Page {p} of {lp}' => 'Page {p} of {lp}',
    'Search Pexels' => "Search Pexels",
	'{n,plural,=1{# picture} other{# pictures}} selected' => '{n,plural,=1{# picture} other{# pictures}} selected',
];


src/PexelsPlugin.php
<?php
namespace convergine\pexels;

use Craft;
use craft\base\Model;
use craft\base\Plugin;
use craft\events\TemplateEvent;
use craft\web\View;
use yii\base\Event;
use convergine\pexels\assets\PexelsAssets;
use convergine\pexels\models\SettingsModel;

class PexelsPlugin extends Plugin
{
	public static $plugin;

	public function init() {

        /* plugin initialization */
		$this->hasCpSection = false;
		$this->hasCpSettings = true;
		parent::init();

        /* register default translations */
        $this->pxDefaultTranslations();
        self::$plugin = $this;

		Event::on(
			View::class,
			View::EVENT_BEFORE_RENDER_PAGE_TEMPLATE,
			function (TemplateEvent $e) {
                /* check conditions before proceeding */
                if (
                    Craft::$app->request->isCpRequest &&
                    !Craft::$app->user->isGuest &&
                    $e->templateMode == 'cp' &&
                    $this->settings->apiKey
                ){
                    /* render templates & construct JS variables */
                    $jsArr = [
                        'pxHolderText' => Craft::t('convergine-pexels', 'placeholderString'),
                        'pxButton' => Craft::$app->view->renderTemplate('convergine-pexels/_components/pexels-button'),
                        'pxModal' => Craft::$app->view->renderTemplate('convergine-pexels/_components/pexels-modal', [
                            'logo' => Craft::$app->assetManager->getPublishedUrl('@convergine/pexels/assets/dist/pexels.svg', true),
                            'placeholderText' => Craft::t('convergine-pexels', 'placeholderString'),
                            ]),
                        'pxLoading' => Craft::$app->view->renderTemplate('convergine-pexels/_components/pexels-loading'),
                        'pxPerPage' => $this->settings->picsPerPage,
                    ];

                    /* generate JS */
                    $scriptContent = '';
                    foreach ($jsArr as $k => $v) {
                        $scriptContent .= ($scriptContent ? ', ' : 'var ') . $k . ' = ' . json_encode($v);
                    }

                    /* register script */
                    Craft::$app->view->registerScript($scriptContent);
                    /* prepare & register CSS and JS */
                    $prepCss = Craft::$app->assetManager->getPublishedUrl('@convergine/pexels/assets/dist/pexels.css', true);
                    $prepJs = Craft::$app->assetManager->getPublishedUrl('@convergine/pexels/assets/dist/pexels.js', true);
                    Craft::$app->view->registerCssFile($prepCss);
                    Craft::$app->view->registerJsFile($prepJs);
				}
		
			}
		);
	
	}

	protected function createSettingsModel(): SettingsModel {
		/* plugin settings model */
		return new SettingsModel();
	}

	protected function settingsHtml(): string {
		/* render plugin settings page template */
		return Craft::$app->view->renderTemplate('convergine-pexels/_settings', ['settings' => $this->settings]);
	}

    public static function tempDirectory()
    {
        /* create a sub-folder for Pexels temporary files in the craft/storage directory
         (or if that can't be found, the plugin directory) and return the location */
        $storageDir = Craft::$app->path->storagePath;

        if (is_dir($storageDir)) {
            $storageDir .= DIRECTORY_SEPARATOR . 'pexels';
        } else {
            $storageDir = __DIR__ . DIRECTORY_SEPARATOR . 'cache';
        }

        if (!is_dir($storageDir)) {
            mkdir($storageDir);
        }

        return $storageDir;
    }


    private function pxDefaultTranslations()
    {
        /* register default translations */
        $translations = [
            'Settings',
            'Please Select',
            'pexelsApiNote' => 'Please note, the Pexels API comes with a default rate limit of 200 requests per hour and 20,000 requests per month.<br/> To request a higher limit, you should reach out to Pexels directly (api@pexels.com). If you adhere to Pexels API terms, you may be eligible to receive unlimited requests for free. <br/>',
            'missingApiKey' => 'API key is missing. Add your Pexels API key and click Save button.',
            'missingUploadLocation' => 'Upload location not set. Select location from dropdown and click Save button.',
            'labelApiKey' => 'Pexels API key',
            'instructionsApiKey' => 'Get your API key <a href="{url}" target="_blank">here</a> if you don\'t have one.',
            'labelUploadLocation' => 'Assets Upload Location',
            'instructionsUploadLocation' => 'Select existing Volume or Folder. If none available - create one first.',
            'convergineHelp' => 'Convergine is not affiliated with Pexels.<br/>If you need help, please open issue on github or check out <a href="{url}" target="_blank">docs</a>.',
            'labelPicsPerPage' => 'Photos Per Page',
            'instructionsPicsPerPage' => 'Number of photos to display per page in search results.',
            'labelLangLocale' => 'Search Language',
            'instructionsLangLocale' => 'Select the language in which you will be searching the Pexels\' photos database.',
            'Any Orientation',
            'Minimum Size',
            'Landscape',
            'Portrait',
            'Square',
            'Small (4MP)',
            'Medium (12MP)',
            'Large (24MP)',
            'Predominant Color',
            'Red',
            'Orange',
            'Yellow',
            'Green',
            'Turquoise',
            'Blue',
            'Violet',
            'Pink',
            'Brown',
            'Black',
            'Gray',
            'White',
            'Download Selected',
            'Downloading, please wait.',
            'placeholderString' => 'Use the search box above to search Pexels pictures database.',
            'An error occurred while loading Pexels data',
            '0 photos found for "{query}"',
            'Previous {picPerPage}',
            'Next {picPerPage}',
            'Page {p} of {lp}',
            'Search Pexels',
            '{n,plural,=1{# picture} other{# pictures}} selected',
        ];

        Craft::$app->view->registerTranslations('convergine-pexels', $translations);
    }

}